File: agenkit/evaluation/optimizer.py
"""
Automated Optimization Framework

This module provides intelligent optimization of agent configurations, prompts,
and hyperparameters using Bayesian optimization, genetic algorithms, and other
search strategies.

Classes:
    Optimizer: Base class for optimization algorithms
    OptimizationResult: Results from optimization run
    SearchSpace: Definition of parameter search space
    RandomSearchOptimizer: Baseline random search
    BayesianOptimizer: Bayesian optimization with Gaussian Process

Example:
    >>> from agenkit.evaluation import BayesianOptimizer
    >>>
    >>> # Define search space
    >>> search_space = {
    ...     "temperature": (0.0, 1.0),
    ...     "top_p": (0.0, 1.0),
    ... }
    >>>
    >>> # Create optimizer
    >>> optimizer = BayesianOptimizer(
    ...     agent_factory=lambda config: MyAgent(**config),
    ...     search_space=search_space,
    ...     objective="accuracy"
    ... )
    >>>
    >>> # Run optimization
    >>> result = await optimizer.optimize(test_cases, n_iterations=50)
    >>> print(f"Best config: {result.best_config}")
    >>> print(f"Best score: {result.best_score:.3f}")
"""

import random
from abc import ABC, abstractmethod
from collections.abc import Callable
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any


@dataclass
class SearchSpace:
    """
    Definition of parameter search space for optimization.

    Supports continuous, discrete, integer, and categorical parameters
    with optional conditional dependencies.

    Example:
        >>> space = SearchSpace()
        >>> space.add_continuous("temperature", 0.0, 1.0)
        >>> space.add_discrete("max_tokens", [128, 256, 512])
        >>> space.add_categorical("model", ["gpt-4", "claude-3"])
    """

    parameters: dict[str, dict[str, Any]] = field(default_factory=dict)

    def add_continuous(self, name: str, low: float, high: float) -> None:
        """Add continuous parameter with range [low, high]."""
        self.parameters[name] = {"type": "continuous", "low": low, "high": high}

    def add_discrete(self, name: str, values: list[int | float]) -> None:
        """Add discrete parameter with specific values."""
        self.parameters[name] = {"type": "discrete", "values": values}

    def add_integer(self, name: str, low: int, high: int) -> None:
        """Add integer parameter with range [low, high]."""
        self.parameters[name] = {"type": "integer", "low": low, "high": high}

    def add_categorical(self, name: str, values: list[str]) -> None:
        """Add categorical parameter with specific values."""
        self.parameters[name] = {"type": "categorical", "values": values}

    def sample(self) -> dict[str, Any]:
        """Sample random configuration from search space."""
        config = {}
        for name, spec in self.parameters.items():
            if spec["type"] == "continuous":
                config[name] = random.uniform(spec["low"], spec["high"])
            elif spec["type"] == "discrete":
                config[name] = random.choice(spec["values"])
            elif spec["type"] == "integer":
                config[name] = random.randint(spec["low"], spec["high"])
            elif spec["type"] == "categorical":
                config[name] = random.choice(spec["values"])
        return config

    def validate(self, config: dict[str, Any]) -> bool:
        """Validate that configuration is within search space."""
        for name, value in config.items():
            if name not in self.parameters:
                return False

            spec = self.parameters[name]
            if spec["type"] == "continuous":
                if not (spec["low"] <= value <= spec["high"]):
                    return False
            elif spec["type"] == "discrete":
                if value not in spec["values"]:
                    return False
            elif spec["type"] == "integer":
                if not isinstance(value, int) or not (spec["low"] <= value <= spec["high"]):
                    return False
            elif spec["type"] == "categorical":
                if value not in spec["values"]:
                    return False

        return True


@dataclass
class OptimizationResult:
    """
    Results from optimization run.

    Attributes:
        best_config: Best configuration found
        best_score: Best objective score achieved
        history: List of (config, score) tuples from all evaluations
        n_iterations: Number of iterations performed
        start_time: Optimization start timestamp
        end_time: Optimization end timestamp
        metadata: Additional metadata
    """

    best_config: dict[str, Any]
    best_score: float
    history: list[tuple[dict[str, Any], float]]
    n_iterations: int
    start_time: str
    end_time: str
    metadata: dict[str, Any] = field(default_factory=dict)

    @property
    def duration_seconds(self) -> float:
        """Duration of optimization in seconds."""
        start = datetime.fromisoformat(self.start_time)
        end = datetime.fromisoformat(self.end_time)
        return (end - start).total_seconds()

    def get_improvement(self) -> float:
        """Improvement from initial to best score."""
        if not self.history:
            return 0.0
        initial_score = self.history[0][1]
        if initial_score == 0:
            return 0.0
        return ((self.best_score - initial_score) / abs(initial_score)) * 100

    def to_dict(self) -> dict[str, Any]:
        """Convert to dictionary."""
        return {
            "best_config": self.best_config,
            "best_score": self.best_score,
            "n_iterations": self.n_iterations,
            "improvement_percent": self.get_improvement(),
            "duration_seconds": self.duration_seconds,
            "start_time": self.start_time,
            "end_time": self.end_time,
            "metadata": self.metadata,
        }


class Optimizer(ABC):
    """
    Base class for optimization algorithms.

    Subclasses should implement the optimize() method to perform
    intelligent search over the configuration space.
    """

    def __init__(
        self,
        agent_factory: Callable[[dict[str, Any]], Any],
        search_space: SearchSpace | dict[str, Any],
        objective: str | Callable,
        maximize: bool = True,
    ):
        """
        Initialize optimizer.

        Args:
            agent_factory: Function that creates agent from config
            search_space: SearchSpace or dict defining parameter space
            objective: Metric name or custom objective function
            maximize: Whether to maximize (True) or minimize (False) objective
        """
        self.agent_factory = agent_factory

        # Convert dict to SearchSpace if needed
        if isinstance(search_space, dict):
            space = SearchSpace()
            for name, spec in search_space.items():
                if isinstance(spec, tuple) and len(spec) == 2:
                    # Continuous range
                    space.add_continuous(name, spec[0], spec[1])
                elif isinstance(spec, list):
                    # Discrete/categorical values
                    if all(isinstance(v, (int, float)) for v in spec):
                        space.add_discrete(name, spec)
                    else:
                        space.add_categorical(name, spec)
            self.search_space = space
        else:
            self.search_space = search_space

        self.objective = objective
        self.maximize = maximize
        self.history: list[tuple[dict[str, Any], float]] = []

    @abstractmethod
    async def optimize(
        self, test_cases: list[dict[str, Any]], n_iterations: int, **kwargs: Any
    ) -> OptimizationResult:
        """
        Run optimization.

        Args:
            test_cases: Test cases for evaluation
            n_iterations: Number of iterations to run
            **kwargs: Additional algorithm-specific parameters

        Returns:
            OptimizationResult with best configuration and history
        """
        pass

    async def evaluate_config(
        self, config: dict[str, Any], test_cases: list[dict[str, Any]]
    ) -> float:
        """
        Evaluate a configuration on test cases.

        Args:
            config: Configuration to evaluate
            test_cases: Test cases for evaluation

        Returns:
            Objective score (higher is better if maximize=True)
        """
        # Create agent with config
        agent = self.agent_factory(config)

        # Evaluate on test cases
        from agenkit.evaluation.core import Evaluator

        # Use objective as metric
        if isinstance(self.objective, str):
            # Create appropriate metrics
            from agenkit.evaluation.quality_metrics import AccuracyMetric, QualityMetrics

            metrics = []
            if self.objective == "accuracy":
                metrics.append(AccuracyMetric(None, False))
            elif self.objective == "quality_score":
                metrics.append(QualityMetrics(False, "", None))

            # Use named metric
            evaluator = Evaluator(agent, metrics, "opt-session")
            result = await evaluator.evaluate(test_cases, "")

            # Get metric value
            if self.objective == "accuracy" and result.accuracy is not None:
                score = result.accuracy
            elif self.objective == "quality_score" and result.quality_score is not None:
                score = result.quality_score
            elif self.objective == "latency_ms" and result.avg_latency_ms is not None:
                score = result.avg_latency_ms
                # Invert for latency (lower is better)
                if self.maximize:
                    score = -score
            else:
                # Default to accuracy
                score = result.accuracy if result.accuracy is not None else 0.0
        else:
            # Use custom objective function
            score = await self.objective(agent, test_cases)

        return score if self.maximize else -score


class RandomSearchOptimizer(Optimizer):
    """
    Baseline random search optimizer.

    Randomly samples configurations from the search space and
    evaluates them. Useful as a baseline for comparison.

    Example:
        >>> optimizer = RandomSearchOptimizer(
        ...     agent_factory=lambda c: MyAgent(**c),
        ...     search_space={"temperature": (0.0, 1.0)},
        ...     objective="accuracy"
        ... )
        >>> result = await optimizer.optimize(test_cases, n_iterations=20)
    """

    async def optimize(
        self, test_cases: list[dict[str, Any]], n_iterations: int, **kwargs: Any
    ) -> OptimizationResult:
        """Run random search optimization."""
        start_time = datetime.now(timezone.utc).isoformat()
        self.history = []

        best_config: dict[str, Any] | None = None
        best_score = float("-inf")

        for _ in range(n_iterations):
            # Sample random configuration
            config = self.search_space.sample()

            # Evaluate
            score = await self.evaluate_config(config, test_cases)
            self.history.append((config.copy(), score))

            # Update best
            if score > best_score:
                best_score = score
                best_config = config.copy()

        end_time = datetime.now(timezone.utc).isoformat()

        if best_config is None:
            best_config = self.search_space.sample()

        return OptimizationResult(
            best_config=best_config,
            best_score=best_score,
            history=self.history,
            n_iterations=n_iterations,
            start_time=start_time,
            end_time=end_time,
            metadata={"algorithm": "random_search"},
        )

File: agenkit-ts/examples/src/evaluation/optimizer.js
"use strict";
/**
 * Automated Optimization Framework
 *
 * This module provides intelligent optimization of agent configurations, prompts,
 * and hyperparameters using Bayesian optimization, genetic algorithms, and other
 * search strategies.
 *
 * Example:
 * ```typescript
 * const searchSpace = new SearchSpace();
 * searchSpace.addContinuous('temperature', 0.0, 1.0);
 * searchSpace.addContinuous('top_p', 0.0, 1.0);
 *
 * const optimizer = new RandomSearchOptimizer(
 *   (config) => new MyAgent(config),
 *   searchSpace,
 *   'accuracy'
 * );
 *
 * const result = await optimizer.optimize(testCases, 50);
 * console.log(`Best config: ${JSON.stringify(result.bestConfig)}`);
 * console.log(`Best score: ${result.bestScore.toFixed(3)}`);
 * ```
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.RandomSearchOptimizer = exports.Optimizer = exports.SearchSpace = void 0;
exports.getOptimizationDuration = getOptimizationDuration;
exports.getOptimizationImprovement = getOptimizationImprovement;
exports.optimizationResultToDict = optimizationResultToDict;
const core_1 = require("./core");
const quality_metrics_1 = require("./quality-metrics");
/**
 * Definition of parameter search space for optimization.
 *
 * Supports continuous, discrete, integer, and categorical parameters.
 */
class SearchSpace {
    constructor() {
        this.parameters = new Map();
    }
    /**
     * Add continuous parameter with range [low, high].
     *
     * @param name Parameter name
     * @param low Minimum value
     * @param high Maximum value
     */
    addContinuous(name, low, high) {
        this.parameters.set(name, { type: 'continuous', low, high });
    }
    /**
     * Add discrete parameter with specific values.
     *
     * @param name Parameter name
     * @param values Valid values for parameter
     */
    addDiscrete(name, values) {
        this.parameters.set(name, { type: 'discrete', values });
    }
    /**
     * Add integer parameter with range [low, high].
     *
     * @param name Parameter name
     * @param low Minimum value
     * @param high Maximum value
     */
    addInteger(name, low, high) {
        this.parameters.set(name, { type: 'integer', low, high });
    }
    /**
     * Add categorical parameter with specific values.
     *
     * @param name Parameter name
     * @param values Valid values for parameter
     */
    addCategorical(name, values) {
        this.parameters.set(name, { type: 'categorical', values });
    }
    /**
     * Sample random configuration from search space.
     *
     * @returns Random configuration
     */
    sample() {
        const config = {};
        for (const [name, spec] of this.parameters.entries()) {
            if (spec.type === 'continuous') {
                const low = spec.low;
                const high = spec.high;
                config[name] = Math.random() * (high - low) + low;
            }
            else if (spec.type === 'discrete' || spec.type === 'categorical') {
                const values = spec.values;
                config[name] = values[Math.floor(Math.random() * values.length)];
            }
            else if (spec.type === 'integer') {
                const low = spec.low;
                const high = spec.high;
                config[name] = Math.floor(Math.random() * (high - low + 1)) + low;
            }
        }
        return config;
    }
    /**
     * Validate that configuration is within search space.
     *
     * @param config Configuration to validate
     * @returns True if configuration is valid
     */
    validate(config) {
        for (const [name, value] of Object.entries(config)) {
            const spec = this.parameters.get(name);
            if (!spec) {
                return false;
            }
            if (spec.type === 'continuous') {
                if (typeof value !== 'number' || value < spec.low || value > spec.high) {
                    return false;
                }
            }
            else if (spec.type === 'discrete') {
                if (!spec.values.includes(value)) {
                    return false;
                }
            }
            else if (spec.type === 'integer') {
                if (typeof value !== 'number' ||
                    !Number.isInteger(value) ||
                    value < spec.low ||
                    value > spec.high) {
                    return false;
                }
            }
            else if (spec.type === 'categorical') {
                if (!spec.values.includes(value)) {
                    return false;
                }
            }
        }
        return true;
    }
    /**
     * Get parameter specification.
     *
     * @param name Parameter name
     * @returns Parameter specification if exists
     */
    getParameter(name) {
        return this.parameters.get(name);
    }
    /**
     * Get all parameter names.
     *
     * @returns Array of parameter names
     */
    getParameterNames() {
        return Array.from(this.parameters.keys());
    }
    /**
     * Get number of parameters.
     *
     * @returns Number of parameters
     */
    size() {
        return this.parameters.size;
    }
}
exports.SearchSpace = SearchSpace;
/**
 * Get duration of optimization in seconds.
 *
 * @param result Optimization result
 * @returns Duration in seconds
 */
function getOptimizationDuration(result) {
    return (result.endTime.getTime() - result.startTime.getTime()) / 1000;
}
/**
 * Get improvement from initial to best score.
 *
 * @param result Optimization result
 * @returns Improvement percentage
 */
function getOptimizationImprovement(result) {
    if (result.history.length === 0) {
        return 0;
    }
    const initialScore = result.history[0][1];
    if (initialScore === 0) {
        return 0;
    }
    return ((result.bestScore - initialScore) / Math.abs(initialScore)) * 100;
}
/**
 * Convert optimization result to plain object.
 *
 * @param result Optimization result
 * @returns Plain object representation
 */
function optimizationResultToDict(result) {
    return {
        best_config: result.bestConfig,
        best_score: result.bestScore,
        n_iterations: result.nIterations,
        improvement_percent: getOptimizationImprovement(result),
        duration_seconds: getOptimizationDuration(result),
        start_time: result.startTime.toISOString(),
        end_time: result.endTime.toISOString(),
        metadata: result.metadata,
    };
}
/**
 * Base class for optimization algorithms.
 *
 * Subclasses should implement the optimize() method to perform
 * intelligent search over the configuration space.
 */
class Optimizer {
    /**
     * Create optimizer.
     *
     * @param agentFactory Function that creates agent from config
     * @param searchSpace SearchSpace defining parameter space
     * @param objective Metric name or custom objective function
     * @param maximize Whether to maximize (true) or minimize (false) objective
     */
    constructor(agentFactory, searchSpace, objective, maximize = true) {
        this.history = [];
        this.agentFactory = agentFactory;
        this.searchSpace = searchSpace;
        this.objective = objective;
        this.maximize = maximize;
    }
    /**
     * Evaluate a configuration on test cases.
     *
     * @param config Configuration to evaluate
     * @param testCases Test cases for evaluation
     * @returns Objective score (higher is better if maximize=true)
     */
    async evaluateConfig(config, testCases) {
        // Create agent with config
        const agent = this.agentFactory(config);
        let score;
        if (typeof this.objective === 'string') {
            // Use named metric
            const metrics = [];
            if (this.objective === 'accuracy') {
                metrics.push(new quality_metrics_1.AccuracyMetric());
            }
            const evaluator = new core_1.Evaluator(agent, metrics, 'opt-session');
            const result = await evaluator.evaluate(testCases);
            // Get metric value
            if (this.objective === 'accuracy' && result.accuracy !== undefined) {
                score = result.accuracy;
            }
            else if (this.objective === 'quality_score' && result.qualityScore !== undefined) {
                score = result.qualityScore;
            }
            else if (this.objective === 'latency_ms' && result.avgLatencyMs !== undefined) {
                score = result.avgLatencyMs;
                // Invert for latency (lower is better)
                if (this.maximize) {
                    score = -score;
                }
            }
            else {
                // Default to accuracy
                score = result.accuracy ?? 0;
            }
        }
        else {
            // Use custom objective function
            score = await this.objective(agent, testCases);
        }
        return this.maximize ? score : -score;
    }
    /**
     * Get optimization history.
     *
     * @returns History of (config, score) pairs
     */
    getHistory() {
        return [...this.history];
    }
}
exports.Optimizer = Optimizer;
/**
 * Baseline random search optimizer.
 *
 * Randomly samples configurations from the search space and
 * evaluates them. Useful as a baseline for comparison.
 */
class RandomSearchOptimizer extends Optimizer {
    /**
     * Run random search optimization.
     *
     * @param testCases Test cases for evaluation
     * @param nIterations Number of iterations to run
     * @returns OptimizationResult with best configuration and history
     */
    async optimize(testCases, nIterations) {
        const startTime = new Date();
        this.history = [];
        let bestConfig = null;
        let bestScore = -Infinity;
        for (let i = 0; i < nIterations; i++) {
            // Sample random configuration
            const config = this.searchSpace.sample();
            // Evaluate
            const score = await this.evaluateConfig(config, testCases);
            this.history.push([{ ...config }, score]);
            // Update best
            if (score > bestScore) {
                bestScore = score;
                bestConfig = { ...config };
            }
        }
        const endTime = new Date();
        // Fallback if no valid config found
        if (bestConfig === null) {
            bestConfig = this.searchSpace.sample();
        }
        return {
            bestConfig,
            bestScore,
            history: this.history,
            nIterations,
            startTime,
            endTime,
            metadata: { algorithm: 'random_search' },
        };
    }
}
exports.RandomSearchOptimizer = RandomSearchOptimizer;
